content/blog/2022/2022-08-03-feature-flags-in-nodejs-backend-server-with-launchdarkly.md

```diff
@@ -0,0 +1,720 @@
+
+```javascript
+import { format } from 'date-fns';
+import padEnd from 'lodash/padEnd.js';
+import capitalize from 'lodash/capitalize.js';
+
+const LEVELS = { debug: 10, log: 20, warn: 30, error: 40 };
+let LD_LOG_LEVEL = LEVELS['debug'];
+
+class Logger {
+  constructor(module) {
+    this.module = module ? module : '';
+
+    this.debug = this.debug.bind(this);
+    this.log = this.log.bind(this);
+    this.warn = this.warn.bind(this);
+    this.error = this.error.bind(this);
+    this.consoleWriter = this.consoleWriter.bind(this);
+  }
+
+  static setLogLevel(level) {
+    LD_LOG_LEVEL = LEVELS[level];
+  }
+
+  static get(module) {
+    return new Logger(module);
+  }
+
+  consoleWriter(level, message, context = '') {
+    if (LEVELS[level] >= LD_LOG_LEVEL) {
+      const dateTime = format(new Date(), 'MM-dd-yyyy HH:mm:ss:SSS');
+      const formattedLevel = padEnd(capitalize(level), 5);
+      const formattedMessage = `${dateTime} ${formattedLevel} [${
+        this.module
+      }] ${message}`;
+      console[level](formattedMessage, context);
+    }
+  }
+
+  debug(message, context) {
+    this.consoleWriter('debug', message, context);
+  }
+
+  log(message, context) {
+    this.consoleWriter('log', message, context);
+  }
+
+  warn(message, context) {
+    this.consoleWriter('warn', message, context);
+  }
+
+  error(message, context) {
+    this.consoleWriter('error', message, context);
+  }
+}
+
+export default Logger;
+```
+
+Then we will define a flag in LaunchDarkly with the name *“backend-log-level”* where we can add a default variation as *“debug”*. We can then change it later to whatever we need.
+
+{{% image alt="Backend Log Level Feature Flag" src="images/posts/nodejs-backend-launchdarkly/Simple_Log_Level.png" %}}
+
+Next we can define a similar kind of code that we did it earlier but we will subscribe to the log level changes before we initiate the `express` app:
+
+```javascript
+import util from 'util';
+import express from 'express';
+import LaunchDarkly from 'launchdarkly-node-server-sdk';
+import Logger from './logger.js';
+
+const PORT = 5000;
+const app = express();
+const simpleLogger = new Logger('Backend');
+
+const LD_SDK_KEY = 'sdk-d2432dc7-e56a-458b-9f93-0361af47d578';
+const LOG_LEVEL_FLAG_KEY = 'backend-log-level';
+const buyerKey = 'mob-b9d6d4d4-4300-46fa-9b13-d9eac89f9794';
+const client = LaunchDarkly.init(LD_SDK_KEY);
+client.waitForInitialization();
+const asyncGetFlag = util.promisify(client.variation);
+
+const subscribeToChanges = () => {
+  client.on(`update:${LOG_LEVEL_FLAG_KEY}`, (_, newValue) => {
+    const {
+      fallthrough: { variation },
+      variations
+    } = newValue;
+    const newLogLevel = variations[variation];
+    console.log(`${LOG_LEVEL_FLAG_KEY} updated to '${newLogLevel}'`);
+    Logger.setLogLevel(newLogLevel);
+  });
+};
+
+client.once('ready', async () => {
+  const user = { key: buyerKey, anonymous: true };
```

Comment:
Do we really want to make the log level specific to a user? That would make sense if we get the user who made the request, but the user is hard-coded here, so it doesn't make sense. 

It would be simpler if we just did not send user information to LaunchDarkly (i.e. send an empty user or no user at all to the `asyncGetFlag()` function. That should work, shouldn't it?